src/mappings/entities/mapAsset.ts

```diff
@@ -0,0 +1,521 @@
+      }).save();
+    }),
+  );
+};
+
+const handleAffirmInstruction = async (params: Record<string, any>) => {
+  const { instructionId } = params;
+  const settlementLegs = await getSettlementLegs(instructionId);
+  await Promise.all(
+    settlementLegs.map(async (leg) => {
+      const asset = await getAsset(leg.ticker);
+      const settlementAmount = new BigNumber(leg.amount);
+      const currentFromAmount = getHolderAmount(leg.from, asset.holders);
+      const currentToAmount = getHolderAmount(leg.to, asset.holders);
+      const otherHolders = asset.holders.filter(
+        (h) => ![leg.from, leg.to].includes(h.did),
+      );
+      asset.holders = [
+        ...otherHolders,
+        {
+          did: leg.from,
+          amount: currentFromAmount.minus(settlementAmount).toString(),
+        },
+        {
+          did: leg.to,
+          amount: currentToAmount.plus(settlementAmount).toString(),
+        },
+      ];
+      asset.totalTransfers = new BigNumber(asset.totalTransfers)
+        .plus(new BigNumber(1))
+        .toString();
+      await asset.save();
+      await Settlement.remove(leg.id);
+    }),
+  );
+};
+
+const handleRejectInstruction = async (params: Record<string, any>) => {
+  const { instructionId } = params;
+  const settlementLegs = await getSettlementLegs(instructionId);
+  await Promise.all(
+    settlementLegs.map(async (leg) => {
+      await Settlement.remove(leg.id);
+    }),
+  );
+};
+// #endregion
+
+// #region ModuleIdEnum.Identity
+const handleAddAuthorization = async (
+  params: Record<string, any>,
+  extrinsic: any,
+) => {
+  const { target: targetData, authorizationData } = params;
+  const id = extrinsic.events[0].event.data[3].toString();
+  const source = extrinsic.events[0].event.data[0].toString();
+  const target = targetData.Identity.toString();
+  const type = Object.keys(authorizationData)[0];
+  if (type === 'TransferAssetOwnership') {
+    const ticker = authorizationData[type].toString();
+    await Authorization.create({
+      id,
+      source,
+      target,
+      type,
+      ticker,
+    }).save();
+  }
+  if (type === 'BecomeAgent') {
+    const ticker = authorizationData[type].col1.toString();
+    const group = Object.keys(authorizationData[type].col2)[0];
+    if (group === 'Full') {
+      await Authorization.create({
+        id,
+        source,
+        target,
+        type,
+        ticker,
+        data: JSON.stringify({ group }),
+      }).save();
+    }
+  }
+};
+
+const handleRemoveAuthorization = async (params: Record<string, any>) => {
+  const { authId } = params;
+  await Authorization.remove(authId);
+};
+// #endregion
+
+// #region ModuleIdEnum.Externalagents
+const handleAcceptBecomeAgent = async (params: Record<string, any>) => {
+  const { authId } = params;
+  const authorization = await getAuthorization(authId);
+  const authorizationData = JSON.parse(authorization.data || '{}');
+  if (!authorization.ticker || authorizationData.group !== 'Full') {
```

Comment:
I think `authorization` or `authorization.data` should have a `type` field to check instead of duck typing on the fields.

If not, can you comment what authorizations you are checking for.

Reply:
`authorization` does have a `type` field. In the `handleAcceptBecomeAgent` function, and for now, we only care about authorization to become full agent, so we're checking for that. Since each authorization type has a different set of params, they're being stored as a JSON string in the `data` field.